### DISCUSSION TITLE
Installation of Monkeytype with local user account creation or signup

### DISCUSSION THREAD

**Discussion by albert-bn** [monkeytypegame/monkeytype#3961] [Questions] [answered]
I want to self-host Monkeytype on my Local Area Network. I have followed the [written instructions](https://github.com/monkeytypegame/monkeytype/blob/master/CONTRIBUTING_ADVANCED.md) on how to do so. I understand that backend installation (i.e. Firebase and Mongodb) is required if I want to have user account creation system.

I have no issue with the frontend installation and it works as expected. However, I just couldn't get the backend to work (i.e. local account creation). I have read through the guides and redo the installation several times but no success.

My setup is:
- Ubuntu Server 22.04 (Virtual Machine)
- Mongodb 4.4.18 (in docker-compose, the image version is 5.0.8 but changed to v4 due to no AVX support)

Am I missing something?

Is local account creation system supported?

When I try to sign up a new account, I realised that it is trying to reach https://api.monkeytype.com/users/signup in the console which make me wonder if account can be created locally.

![image](https://user-images.githubusercontent.com/47083111/215312936-380de559-bb74-4e5a-9176-890590f944aa.png)

**Comment by Miodec** [comment, marked as answer]
You need to check your database and look up that error code.

**Reply by albert-bn** [reply]
Sorry that I'm new to Node.js and MongoDB. Do you mind if you can give a few more pointers?

1) How can I check if 'frontend' is connected to 'backend'? I have both docker-compose up and running.

2) When I check in MongoDB Compass, there is no 'monkeytype' database. Do I need to create the database manually?

Really appreciate your help.

**Reply by Miodec** [reply]
If the frontend is NOT connected to the backend you will see this message: 
![image](https://user-images.githubusercontent.com/13181393/215462265-0e942f60-671a-4d44-a081-08f5f0ab9057.png)

But if you're getting the "monkeys dropped our bananas" message then the backend server is definitely running.

I *think* the database should create itself once you try to add some data by signing up.

**Reply by albert-bn** [reply]
If I access Monkeytype via http://localhost:3000, it shows the "Backend server not running" message.
![image](https://user-images.githubusercontent.com/47083111/215498678-9c512fe0-62d3-4762-9e69-634b7721a3a4.png)

Somehow, if access with my machine IP address http://192.168.57.73:3000, there is no warning message.
![image](https://user-images.githubusercontent.com/47083111/215499087-11d13160-340e-401b-9f36-aa3e4c940f06.png)

Either way, both addresses do not let me create user account. Does it mean that mongodb is not connected?

Where else should I check?

Thanks again.

**Reply by Miodec** [reply]
You need to use localhost. That would mean the backend server is not running. How are you running the project?

(if you use your local ip the frontend will not recognise its running locally and will send requests to the live API server)

**Reply by albert-bn** [reply]
Noted on the localhost address to process the request locally.

As for my project setup. I used `Ubuntu 22.04` and I followed the instructions given at https://github.com/monkeytypegame/monkeytype/blob/master/CONTRIBUTING_ADVANCED.md

1. Installed `Node.js 16.13.2` with `NPM 8.1.2`
2. Created a Firebase project as per instruction. Saved the `serviceAccountKey.json` to `backend/src/credentials/` folder, and `.firebaserc` in `frontend/` folder
3. `firebase-config.ts` saved in `frontend/src/ts/constants` . Do you want to see the firebase-config.ts file?
4. `.env` file in `backend/` . I did not edit anything in .env file (everything is as default from example.env)
5. I run both '`frontend`' and '`backend`' with `docker-compose` as recommended in written instructions.

![image](https://user-images.githubusercontent.com/47083111/215510348-0ac1bd96-fd0e-4f8c-b543-9038b42ba4c3.png)


Another side note is that all the files ownership is under 'root' user, hence, everything run with 'root' user.

**Reply by Miodec** [reply]
Right, the instructions are incorrect. The backend compose only runs the mongo and redis servers. You gonna need to run `npm run dev` in project root.

**Reply by albert-bn** [reply]
It works!

One last question. When I deploy the project to production, obviously I can't use localhost from other LAN machine to access.

During the deployment, does it mean that I have to change the value of `PROD_SERVER_HOST` in `frontend/src/ts/ape/index.ts` to my ip address or hostname?

![image](https://user-images.githubusercontent.com/47083111/215515642-171c01e1-f848-4efb-9522-6d5bc55cd42b.png)

**Reply by Miodec** [reply]
Yeap. It has to point to the API server you want to use.

**Reply by albert-bn** [reply]
That's cool. You have been such a great help, and being patience to newbie like me.

**Reply by Juliaria08** [reply]
@Miodec Would it be possible to deploy the backend authentication service without using firebase. And if that wouldn't be possible. Would it be possible for me make a custom backend that doesn't use firebase (or maybe even just modify the frontend code to use mysql or some other db)?

**Reply by Miodec** [reply]
Not possible. Feel free to make that change if you want, but its gonna be very large and probably wont be merged to master

**Reply by Juliaria08** [reply]
> Not possible. Feel free to make that change if you want, but its gonna be very large and probably wont be merged to master

Thanks for the help.
